File: src/games/sudoku/gameplay/analysis/techniqueModules/chains.ts
import type { SudokuDigit } from '../../../types';
import type { SudokuCanonicalMove } from '../moves';
import { popcount } from '../bitmask';
import { arePeerIndexes, buildCandidateEliminationMove, getCellCandidatesByIndex } from '../techniqueHelpers';
import type { SudokuTechniqueDispatcher } from '../techniqueModuleTypes';

const MAX_CHAIN_LENGTH = 6;

export const chainsTechnique: SudokuTechniqueDispatcher = {
  technique: 'chains',
  tier: 'expert',
  findMove(state) {
    const bivalueCells = state.board
      .map((value, index) => ({ value, index }))
      .filter(({ value, index }) => value === 0 && popcount[state.candidateMask[index]] === 2)
      .map(({ index }) => ({ index, candidates: getCellCandidatesByIndex(state, index) }));

    const search = (
      startIndex: number,
      currentIndex: number,
      targetDigit: SudokuDigit,
      sharedDigit: SudokuDigit,
      path: number[],
    ): SudokuCanonicalMove | null => {
      if (path.length >= MAX_CHAIN_LENGTH) {
        return null;
      }

      const nextMoves = bivalueCells
        .filter((cell) => !path.includes(cell.index) && arePeerIndexes(cell.index, currentIndex))
        .filter((cell) => cell.candidates.includes(sharedDigit));

      for (const nextCell of nextMoves) {
        const nextDigit = nextCell.candidates.find((digit) => digit !== sharedDigit);
        if (!nextDigit) {
          continue;
        }

        const chainLength = path.length + 1;

        if (nextCell.candidates.includes(targetDigit) && path.length >= 3) {
          const move = buildCandidateEliminationMove({
            technique: 'chains',
            eliminations: state.board
              .map((value, index) => ({ value, index }))
              .filter(({ value, index }) => value === 0 && ![...path, nextCell.index].includes(index))
              .filter(({ index }) => arePeerIndexes(index, startIndex) && arePeerIndexes(index, nextCell.index))
              .filter(({ index }) => getCellCandidatesByIndex(state, index).includes(targetDigit))
              .map(({ index }) => ({ index, digit: targetDigit })),
            evidenceCells: [...path, nextCell.index],
            complexity: chainLength,
          });
          if (move) {
            return move;
          }
        }

        const nested = search(startIndex, nextCell.index, targetDigit, nextDigit, [...path, nextCell.index]);
        if (nested) {
          return nested;
        }
      }

      return null;
    };

    for (const startCell of bivalueCells) {
      const [firstDigit, secondDigit] = startCell.candidates;
      const firstMove = search(startCell.index, startCell.index, firstDigit, secondDigit, [startCell.index]);
      if (firstMove) {
        return firstMove;
      }

      const secondMove = search(startCell.index, startCell.index, secondDigit, firstDigit, [startCell.index]);
      if (secondMove) {
        return secondMove;
      }
    }

    return null;
  },
};

File: engine/src/main/kotlin/com/quietgrid/engine/sudoku/techniques/Chains.kt
package com.quietgrid.engine.sudoku.techniques

import com.quietgrid.engine.core.Difficulty
import com.quietgrid.engine.sudoku.*

object ChainsTechnique : SudokuTechniqueDispatcher {
    override val technique = SudokuTechnique.CHAINS
    override val tier = Difficulty.EXPERT

    private const val MAX_CHAIN_LENGTH = 6

    private data class Cell(val index: Int, val candidates: List<Int>)

    override fun findMove(state: SudokuBitmaskState): SudokuCanonicalMove? {
        val bivalueCells = state.board.indices.filter { state.board[it] == 0 && popcount[state.candidateMask[it]] == 2 }
            .map { Cell(it, getCellCandidatesByIndex(state, it)) }

        fun search(startIndex: Int, currentIndex: Int, targetDigit: Int, sharedDigit: Int, path: List<Int>): SudokuCanonicalMove? {
            if (path.size >= MAX_CHAIN_LENGTH) return null

            val nextMoves = bivalueCells.filter { it.index !in path && arePeerIndexes(it.index, currentIndex) && it.candidates.contains(sharedDigit) }

            for (nextCell in nextMoves) {
                val nextDigit = nextCell.candidates.firstOrNull { it != sharedDigit } ?: continue
                val chainLength = path.size + 1

                if (nextCell.candidates.contains(targetDigit) && path.size >= 3) {
                    val fullPath = path + nextCell.index
                    val eliminations = state.board.indices
                        .filter { state.board[it] == 0 && it !in fullPath }
                        .filter { arePeerIndexes(it, startIndex) && arePeerIndexes(it, nextCell.index) }
                        .filter { getCellCandidatesByIndex(state, it).contains(targetDigit) }
                        .map { it to targetDigit }
                    val move = buildCandidateEliminationMove(SudokuTechnique.CHAINS, eliminations, fullPath, null, chainLength)
                    if (move != null) return move
                }

                val nested = search(startIndex, nextCell.index, targetDigit, nextDigit, path + nextCell.index)
                if (nested != null) return nested
            }
            return null
        }

        for (startCell in bivalueCells) {
            val (firstDigit, secondDigit) = startCell.candidates
            val firstMove = search(startCell.index, startCell.index, firstDigit, secondDigit, listOf(startCell.index))
            if (firstMove != null) return firstMove

            val secondMove = search(startCell.index, startCell.index, secondDigit, firstDigit, listOf(startCell.index))
            if (secondMove != null) return secondMove
        }
        return null
    }
}
